File: web/src/lib/validation.ts
import { Profile } from "./api";

const MONTH_YEAR = /^(0[1-9]|1[0-2])\/\d{4}$/;
const EMAIL = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
const PHONE = /^[\d\s\-+().]{7,20}$/;

export function monthInputToDisplay(value: string): string {
  if (!value) return "";
  if (MONTH_YEAR.test(value)) return value;
  if (/^\d{4}-\d{2}$/.test(value)) {
    const [y, m] = value.split("-");
    return `${m}/${y}`;
  }
  return value;
}

export function monthInputFromDisplay(value: string): string {
  const v = value.trim();
  if (!v) return "";
  if (/^\d{4}-\d{2}$/.test(v)) return v;
  if (MONTH_YEAR.test(v)) {
    const [m, y] = v.split("/");
    return `${y}-${m}`;
  }
  return v;
}

export function validateStep(
  step: number,
  profile: Omit<Profile, "id">,
  jobDescription: string,
  forGenerate = false
): string[] {
  const errors: string[] = [];

  if (step === 0 || forGenerate) {
    if (!profile.full_name.trim()) errors.push("Full name is required");
    if (!profile.email.trim()) errors.push("Email is required");
    else if (!EMAIL.test(profile.email)) errors.push("Enter a valid email");
    if (!profile.phone.trim()) errors.push("Phone is required");
    else if (!PHONE.test(profile.phone.replace(/\s/g, "")))
      errors.push("Enter a valid phone number for your country");
    if (profile.linkedin.trim() && !profile.linkedin.toLowerCase().includes("linkedin.com")) {
      errors.push("LinkedIn should be a linkedin.com URL");
    }
    if (profile.years_experience < 0 || profile.years_experience > 60) {
      errors.push("Years of experience: 0–60");
    }
    if (forGenerate && step !== 0) return errors;
    if (!forGenerate && step === 0) return errors;
  }

  if (step === 1 || forGenerate) {
    if (forGenerate && profile.experience.length === 0) {
      errors.push("Add at least one job");
    }
    profile.experience.forEach((job, i) => {
      const n = i + 1;
      if (!forGenerate && !job.company && !job.title) return;
      if (!job.company.trim()) errors.push(`Job ${n}: Company required`);
      if (!job.title.trim()) errors.push(`Job ${n}: Title required`);
      if (job.start_date && !MONTH_YEAR.test(job.start_date) && !/^\d{4}-\d{2}$/.test(job.start_date)) {
        errors.push(`Job ${n}: Start date → MM/YYYY`);
      }
      if (forGenerate && !job.start_date.trim()) errors.push(`Job ${n}: Start date required`);
      const end = (job.end_date || "").trim();
      if (
        end &&
        end.toLowerCase() !== "present" &&
        !MONTH_YEAR.test(end) &&
        !/^\d{4}-\d{2}$/.test(end)
      ) {
        errors.push(`Job ${n}: End → Present, MM/YYYY, or leave empty`);
      }
      if (forGenerate) {
        const bullets = job.bullets.filter((b) => b.trim());
        if (!bullets.length) errors.push(`Job ${n}: At least one bullet`);
      }
    });
    if (!forGenerate && step === 1) return errors;
  }

  if (step === 2) {
    profile.education.forEach((edu, i) => {
      if (!edu.institution && !edu.degree) return;
      if (!edu.institution.trim()) errors.push(`Education ${i + 1}: Institution required`);
      if (!edu.degree.trim()) errors.push(`Education ${i + 1}: Degree required`);
    });
    return errors;
  }

  if (step === 3 || forGenerate) {
    if (forGenerate && profile.skills.filter((s) => s.trim()).length < 3) {
      errors.push("Add at least 3 skills");
    }
    if (!forGenerate && step === 3) return errors;
  }

  if (step === 4 || forGenerate) {
    if (forGenerate && jobDescription.trim().length < 50) {
      errors.push("Paste full job description (50+ characters)");
    }
  }

  return errors;
}

File: resume_engine/validation.py
"""Shared validation for profiles, dates, and wizard steps."""

import re
from typing import Any

MONTH_YEAR = re.compile(r"^(0[1-9]|1[0-2])/\d{4}$")
YEAR_ONLY = re.compile(r"^\d{4}$")
EMAIL = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")
PHONE = re.compile(r"^[\d\s\-+().]{7,20}$")


def normalize_month_year(value: str) -> str:
    """Convert YYYY-MM (from month input) or loose input to MM/YYYY."""
    v = value.strip()
    if not v:
        return ""
    if MONTH_YEAR.match(v):
        return v
    if re.match(r"^\d{4}-\d{2}$", v):
        y, m = v.split("-")
        return f"{m}/{y}"
    if YEAR_ONLY.match(v):
        return f"01/{v}"
    return v


def is_valid_month_year(value: str, required: bool = True) -> bool:
    if not value or not value.strip():
        return not required
    v = value.strip()
    if v.lower() == "present":
        return True
    return bool(MONTH_YEAR.match(v) or YEAR_ONLY.match(v) or re.match(r"^\d{4}-\d{2}$", v))


def is_valid_end_date(value: str) -> bool:
    if not value or not value.strip():
        return True
    v = value.strip()
    if v.lower() == "present":
        return True
    return is_valid_month_year(v, required=True)


def is_valid_email(value: str) -> bool:
    return bool(value and EMAIL.match(value.strip()))


def is_valid_phone(value: str, country_code: str = "+1") -> bool:
    from resume_engine.utils.phone import validate_phone as vp

    ok, _ = vp(country_code, value)
    return ok


def is_valid_linkedin(value: str) -> bool:
    if not value or not value.strip():
        return True
    v = value.strip().lower()
    return "linkedin.com" in v or v.startswith("http")


def validate_profile_step(data: dict[str, Any], *, strict: bool = False) -> list[str]:
    errors: list[str] = []
    if not (data.get("full_name") or "").strip():
        errors.append("Full name is required")
    if strict or (data.get("email") or "").strip():
        if not is_valid_email(data.get("email", "")):
            errors.append("Valid email is required (e.g. you@company.com)")
    elif strict:
        errors.append("Email is required")
    cc = data.get("phone_country_code") or "+1"
    if strict or (data.get("phone") or "").strip():
        if not is_valid_phone(data.get("phone", ""), cc):
            errors.append("Valid phone number required for selected country code")
    elif strict:
        errors.append("Phone is required")
    if (data.get("linkedin") or "").strip() and not is_valid_linkedin(data.get("linkedin", "")):
        errors.append("LinkedIn should be a full linkedin.com URL")
    years = data.get("years_experience", 0)
    if years is not None and (years < 0 or years > 60):
        errors.append("Years of experience must be between 0 and 60")
    return errors


def validate_experience_step(experience: list[dict], *, for_generate: bool = False) -> list[str]:
    errors: list[str] = []
    if for_generate and not experience:
        errors.append("Add at least one job before generating")
        return errors
    for i, job in enumerate(experience):
        prefix = f"Job {i + 1}"
        if for_generate or job.get("company") or job.get("title"):
            if not (job.get("company") or "").strip():
                errors.append(f"{prefix}: Company is required")
            if not (job.get("title") or "").strip():
                errors.append(f"{prefix}: Job title is required")
            start = job.get("start_date", "")
            if for_generate and not is_valid_month_year(start, required=True):
                errors.append(f"{prefix}: Start date must be MM/YYYY (e.g. 03/2021)")
            elif start and not is_valid_month_year(start, required=False):
                errors.append(f"{prefix}: Invalid start date format")
            end = (job.get("end_date") or "").strip()
            if end and not is_valid_end_date(end):
                errors.append(f"{prefix}: End date must be Present or MM/YYYY")
            if for_generate:
                bullets = [b for b in (job.get("bullets") or []) if b and str(b).strip()]
                if not bullets:
                    errors.append(f"{prefix}: Add at least one bullet point")
    return errors


def validate_education_step(education: list[dict]) -> list[str]:
    errors: list[str] = []
    for i, edu in enumerate(education):
        has_any = any((edu.get(k) or "").strip() for k in ("institution", "degree", "field"))
        if not has_any:
            continue
        if not (edu.get("institution") or "").strip():
            errors.append(f"Education {i + 1}: Institution is required when adding education")
        if not (edu.get("degree") or "").strip():
            errors.append(f"Education {i + 1}: Degree is required when adding education")
        grad = edu.get("graduation_date") or ""
        if grad and not (is_valid_month_year(grad, False) or YEAR_ONLY.match(grad.strip())):
            errors.append(f"Education {i + 1}: Graduation must be YYYY or MM/YYYY")
    return errors


def _normalize_skills(skills: list[str]) -> list[str]:
    """Split combined entries (commas, semicolons, newlines) into separate skills."""
    import re

    out: list[str] = []
    seen: set[str] = set()
    for item in skills:
        for part in re.split(r"[,;|\n]+", item or ""):
            s = re.sub(r"^[\s•\-*]+", "", part.strip())
            if not s:
                continue
            key = s.lower()
            if key not in seen:
                seen.add(key)
                out.append(s)
    return out


def validate_skills_step(skills: list[str], *, for_generate: bool = False) -> list[str]:
    normalized = _normalize_skills(skills)
    if for_generate and len(normalized) < 3:
        return ["Add at least 3 skills before generating (use commas or new lines between each)"]
    return []


def validate_job_description(jd: str, *, required: bool = False) -> list[str]:
    if required and len((jd or "").strip()) < 50:
        return ["Paste the full job description (at least 50 characters)"]
    return []
